shims/npm

```diff
@@ -51,18 +49,85 @@ should_reshim() {
       ;;
   esac
 
-  if [ "$is_global" ] && [ "$cmd_needs_reshim" ]; then
-    return 0
+  # Implicit return
+  [ "$is_global" ] && [ "$cmd_needs_reshim" ]
+}
+
+resolve_canon_npm() {
+  local npm_location=
+
+  # Try searching in current path (asdf core from 0.7 onwards adds all binaries candidates directories to PATH)
+  # if that doesn't works (when calling the shim directly for example) it tries manually searching the binary in
+  # the installed version provided by "asdf where"
+  npm_location="${ASDF_NODEJS_CANON_NPM_PATH:-$(search_npm_on_current_path || manually_search_npm_bin)}"
+
+  if ! [ "$npm_location" ]; then
+    printf "asdf-nodejs couldn't find a suitable npm executable\n"
+    printf "This is probably a problem with the plugin, please report this issue at https://github.com/asdf-vm/asdf-nodejs/issues\n"
+    exit 1
   fi
 
+  printf "%s\n" "$npm_location"
+}
+
+remove_current_dir_from_path() {
+  local filtered_path= dir= normalized_dir=
+
+  while read -rd : dir; do
+    if [ -d "$dir" ]; then
+      normalized_dir=$(cd "$dir" && pwd -P)
+
+      if [ "$normalized_dir" = "$this_dir" ] || [ "$normalized_dir" = "$asdf_shims_dir" ]; then
+        continue
+      fi
+    fi
+
+    filtered_path+="$dir:"
+  done <<< "$PATH:"
+
+  printf "%s\n" "${filtered_path%:}"
+}
+
+search_npm_on_current_path() {
+  local filtered_path=
+
+  # Tries to prevent recursion by removing the current script and asdf-shim from PATH
+  filtered_path=$(remove_current_dir_from_path)
+
+  PATH="$filtered_path" command -v npm
+}
+
+manually_search_npm_bin() {
+  local installed_dir= bin_directories=
+
+  installed_dir=$(asdf where nodejs)
```

Comment:
I'm not sure this is the best solution here. As a general rule asdf plugins shouldn't invoke asdf because it creates a circular dependency.